File: KV_DS_Assignment_Module20_cococolasales.py
# -*- coding: utf-8 -*-
"""
Created on Fri Jul 17 19:30:09 2020

@author: KavithaV
"""

import pandas as pd
cococola = pd.read_excel("D:/Data science/Module 20 - Forecast and Time series/CocaCola_Sales_Rawdata (1)/CocaCola_Sales_Rawdata (1).xlsx")
import numpy as np
# Data Preparation
cococola["t"] = np.arange(1,43)

cococola["t_squared"] = cococola["t"]*cococola["t"]
cococola.columns
cococola["log_sales"] = np.log(cococola["Sales"])
#Exract quarter from quarter/year column
cococola["Quart"] = cococola["Quarter"].str[0:2]
# Dummy variables for Seasons
months =['Q1','Q2','Q3','Q4'] 

month_dummies = pd.DataFrame(pd.get_dummies(cococola['Quart']))
cococola = pd.concat([cococola,month_dummies],axis = 1)

# Partition the data
cococola.iloc[:, 1].plot() # Timeplot
Train = cococola.head(30)
Test = cococola.tail(12)

####################### L I N E A R ##########################
import statsmodels.formula.api as smf 

linear_model = smf.ols('Sales ~ t', data=Train).fit()
pred_linear =  pd.Series(linear_model.predict(pd.DataFrame(Test['t'])))
rmse_linear = np.sqrt(np.mean((np.array(Test['Sales'])-np.array(pred_linear))**2))
rmse_linear
#Out[21]: 714.0144483281297
##################### Exponential ##############################

Exp = smf.ols('log_sales ~ t', data=Train).fit()
pred_Exp = pd.Series(Exp.predict(pd.DataFrame(Test['t'])))
rmse_Exp = np.sqrt(np.mean((np.array(Test['Sales'])-np.array(np.exp(pred_Exp)))**2))
rmse_Exp
#Out[25]: 552.2821039079295
#################### Quadratic ###############################

Quad = smf.ols('Sales ~ t + t_squared',data=Train).fit()
pred_Quad = pd.Series(Quad.predict(Test[["t","t_squared"]]))
rmse_Quad = np.sqrt(np.mean((np.array(Test['Sales'])-np.array(pred_Quad))**2))
rmse_Quad
#Out[29]: 646.2715428311323
################### Additive seasonality ########################

add_sea = smf.ols('Sales ~ Q1+Q2+Q3+Q4', data=Train).fit()
pred_add_sea = pd.Series(add_sea.predict(Test[['Q1','Q2','Q3','Q4']]))
rmse_add_sea = np.sqrt(np.mean((np.array(Test['Sales'])-np.array(pred_add_sea))**2))
rmse_add_sea
#Out[34]: 1778.0065467941363
################## Multiplicative Seasonality ##################

Mul_sea = smf.ols('log_sales ~ Q1+Q2+Q3+Q4',data = Train).fit()
pred_Mult_sea = pd.Series(Mul_sea.predict(Test))
rmse_Mult_sea = np.sqrt(np.mean((np.array(Test['Sales'])-np.array(np.exp(pred_Mult_sea)))**2))
rmse_Mult_sea
#Out[38]: 1828.9238912138073
################## Additive Seasonality Quadratic Trend ############################

add_sea_Quad = smf.ols('Sales ~ t+t_squared+Q1+Q2+Q3+Q4',data=Train).fit()
pred_add_sea_quad = pd.Series(add_sea_Quad.predict(Test[['Q1','Q2','Q3','Q4','t','t_squared']]))
rmse_add_sea_quad = np.sqrt(np.mean((np.array(Test['Sales'])-np.array(pred_add_sea_quad))**2))
rmse_add_sea_quad 
#Out[42]: 586.0533067871024
################## Multiplicative Seasonality and Exponential ###########

Mul_Add_sea = smf.ols('log_sales ~ t+Q1+Q2+Q3+Q4',data = Train).fit()
pred_Mult_add_sea = pd.Series(Mul_Add_sea.predict(Test))
rmse_Mult_add_sea = np.sqrt(np.mean((np.array(Test['Sales'])-np.array(np.exp(pred_Mult_add_sea)))**2))
rmse_Mult_add_sea 
#Out[47]: 410.24970596078435
################## Testing #######################################

data = {"MODEL":pd.Series(["rmse_linear","rmse_Exp","rmse_Quad","rmse_add_sea","rmse_add_sea_quad","rmse_Mult_sea","rmse_Mult_add_sea"]),"RMSE_Values":pd.Series([rmse_linear,rmse_Exp,rmse_Quad,rmse_add_sea,rmse_add_sea_quad,rmse_Mult_sea,rmse_Mult_add_sea])}
table_rmse=pd.DataFrame(data)
table_rmse

# so rmse_Mult_add_sea has the least value among the models prepared so far 

File: KV_DS_Assignment_Module20_cococolasales.R
# Load the Dataset
library(readr)
library(readxl)
cococola <- read_xlsx("D:/Data science/Module 20 - Forecast and Time series/CocaCola_Sales_Rawdata (1)/CocaCola_Sales_Rawdata (1).xlsx") # read the cococola data
View(cococola) 

windows()
plot(cococola$Sales, type="o")

# Data Preparation
cococola["t"] <- 1:42
View(cococola)

cococola["t_square"] <- cococola["t"] * cococola["t"]
cococola["log_Sales"] <- log(cococola["Sales"])
library(stringr)
cococola["Quarter"] = str_sub(cococola$Quarter,1,2) 

install.packages('fastDummies')
library('fastDummies')
cococola <- dummy_cols(cococola, select_columns = 'Quarter')

# Partition the time series
train <- cococola[1:30, ]
test <- cococola[31:42, ]


########################### LINEAR MODEL #############################
linear_model <- lm(Sales ~ t, data=train)
summary(linear_model)

linear_pred <- data.frame(predict(linear_model,interval='predict', newdata =test))
View(linear_pred)

rmse_linear <- sqrt(mean((test$Sales - linear_pred$fit)^2, na.rm = T))
rmse_linear # 714.0144



######################### Exponential #################################

expo_model <- lm(log_Sales ~ t, data=train)
summary(expo_model)

expo_pred <- data.frame(predict(expo_model, interval='predict', newdata=test))
View(expo_pred)

rmse_expo<-sqrt(mean((test$Sales - exp(expo_pred$fit))^2, na.rm = T))
rmse_expo # 552.2821

######################### Quadratic ####################################

Quad_model <- lm(Sales ~ t + t_square, data=train)
summary(Quad_model)
Quad_pred <- data.frame(predict(Quad_model, interval='predict', newdata=test))

rmse_Quad <- sqrt(mean((test$Sales - Quad_pred$fit)^2, na.rm=T))
rmse_Quad # 646.2715

######################### Additive Seasonality #########################

sea_add_model <- lm(Sales ~ Quarter_Q1+Quarter_Q2+Quarter_Q3+Quarter_Q4, data=train)
summary(sea_add_model)

sea_add_pred <- data.frame(predict(sea_add_model, newdata=test, interval='predict'))

rmse_sea_add <- sqrt(mean((test$Sales - sea_add_pred$fit)^2, na.rm = T))
rmse_sea_add # 1778.007

######################## Multiplicative Seasonality #########################

multi_sea_model <- lm(log_Sales ~ Quarter_Q1+Quarter_Q2+Quarter_Q3+Quarter_Q4, data = train)
summary(multi_sea_model)

multi_sea_pred <- data.frame(predict(multi_sea_model, newdata=test, interval='predict'))

rmse_multi_sea <- sqrt(mean((test$Sales - exp(multi_sea_pred$fit))^2, na.rm = T))
rmse_multi_sea # 1828.924

######################## Additive Seasonality with Linear trend #################

Add_sea_Linear_model <- lm(Sales ~ t+Quarter_Q1+Quarter_Q2+Quarter_Q3+Quarter_Q4,data=train)
summary(Add_sea_Linear_model)
Add_sea_Linear_pred<-data.frame(predict(Add_sea_Linear_model,interval='predict',newdata=test))
rmse_Add_sea_Linear<-sqrt(mean((test$Sales-Add_sea_Linear_pred$fit)^2,na.rm=T))
rmse_Add_sea_Linear # 637.9405

######################## Additive Seasonality with Quadratic Trend #################

Add_sea_Quad_model<-lm(Sales ~ t+t_square+Quarter_Q1+Quarter_Q2+Quarter_Q3+Quarter_Q4,data=train)
summary(Add_sea_Quad_model)

Add_sea_Quad_pred<-data.frame(predict(Add_sea_Quad_model,interval='predict',newdata=test))
rmse_Add_sea_Quad<-sqrt(mean((test$Sales-Add_sea_Quad_pred$fit)^2,na.rm=T))
rmse_Add_sea_Quad # 586.0533

######################## Multiplicative Seasonality Linear trend ##########################

multi_add_sea_model<-lm(log_Sales ~ t+Quarter_Q1+Quarter_Q2+Quarter_Q3+Quarter_Q4,data = train)
summary(multi_add_sea_model) 

multi_add_sea_pred<-data.frame(predict(multi_add_sea_model,newdata=test,interval='predict'))
rmse_multi_add_sea<-sqrt(mean((test$Sales-exp(multi_add_sea_pred$fit))^2,na.rm = T))
rmse_multi_add_sea # 410.2497
###### Preparing table on model and it's RMSE values 

table_rmse <- data.frame(c("rmse_linear","rmse_expo","rmse_Quad","rmse_sea_add","rmse_multi_sea","rmse_Add_sea_Quad","rmse_multi_add_sea"),c(rmse_linear,rmse_expo,rmse_Quad,rmse_sea_add,rmse_multi_sea,rmse_Add_sea_Quad,rmse_multi_add_sea))
View(table_rmse)

colnames(table_rmse) <-c ("model","RMSE")
View(table_rmse)

############### Multiplicative Seasonality Linear trend has least RMSE value############################# Additive seasonality with Quadratic Trend has least RMSE value

new_model <- lm(log_Sales ~ t+Quarter_Q1+Quarter_Q2+Quarter_Q3+Quarter_Q4,data = train)
summary(new_model)

resid <- residuals(new_model)
resid[1:30]

# Predict
library(readr)
new_data <- read_csv(file.choose())
View(new_data)

pred_new <- data.frame(predict(new_model, newdata=new_data, interval = 'predict'))
View(pred_new)
pred_new$fit
 